build.assets/tooling/lib/eventschema/schemagen.go

```diff
@@ -0,0 +1,248 @@
+	name := message.Name()
+	if schema, ok := generator.memo[name]; ok {
+		if !schema.built {
+			return nil, trace.Errorf("circular dependency in the %s", message.Name())
+		}
+		return schema, nil
+	}
+	schema := NewSchema(descriptionFromComment(message.LeadingComments()))
+	generator.memo[name] = schema
+
+	for _, field := range message.Fields {
+		if _, ok := ignoredFields[message.Name()][field.Name()]; ok {
+			continue
+		}
+
+		switch jsonName := field.JSONName(); jsonName {
+		case "-":
+			// We skip this field
+		case "":
+			// The field is an embedded message, we traverse it and copy its
+			// properties to the parent
+			if !field.IsMessage() {
+				return nil, trace.BadParameter("Embedded field is not a message")
+			}
+			embeddedSchema, err := generator.traverseInner(field.TypeMessage())
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			for propName, prop := range embeddedSchema.Properties {
+				schema.Properties[propName] = prop
+			}
+		default:
+			// This is a regular non-embedded field
+			var err error
+			schema.Properties[jsonName], err = generator.prop(field)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+	}
+	schema.built = true
+
+	return schema, nil
+}
+
+func (generator *SchemaGenerator) prop(field *tree.Field) (apiextv1.JSONSchemaProps, error) {
+	prop := apiextv1.JSONSchemaProps{Description: descriptionFromComment(field.LeadingComments())}
+
+	if field.IsRepeated() && !field.IsMap() {
+		prop.Type = "array"
+		prop.Items = &apiextv1.JSONSchemaPropsOrArray{
+			Schema: &apiextv1.JSONSchemaProps{},
+		}
+		if err := generator.singularProp(field, prop.Items.Schema); err != nil {
+			return prop, trace.Wrap(err)
+		}
+	} else {
+		if err := generator.singularProp(field, &prop); err != nil {
+			return prop, trace.Wrap(err)
+		}
+	}
+
+	if field.IsNullable() && (prop.Type == "array" || prop.Type == "object") {
+		prop.Nullable = true
+	}
+
+	// Labels are relying on `utils.Strings`, which can either marshall as an array of strings or a single string
+	// This does not pass Schema validation from the apiserver, to workaround we don't specify type for those fields
+	// and ask Kubernetes to preserve unknown fields.
+	if field.CustomType() == "Labels" {
+		prop.Type = "object"
+		preserveUnknownFields := true
+		prop.AdditionalProperties = &apiextv1.JSONSchemaPropsOrBool{
+			Schema: &apiextv1.JSONSchemaProps{
+				XPreserveUnknownFields: &preserveUnknownFields,
+			},
+		}
+	}
+
+	return prop, nil
+}
+
+func (generator *SchemaGenerator) singularProp(field *tree.Field, prop *apiextv1.JSONSchemaProps) error {
+	switch {
+	case field.IsBool():
+		prop.Type = "boolean"
+	case field.IsString():
+		prop.Type = "string"
+	case field.IsDuration():
+		prop.Type = "string"
+		prop.Format = "duration"
+	case field.IsTime():
+		prop.Type = "string"
+		prop.Format = "date-time"
+	case field.IsInt32() || field.IsUint32() || field.IsEnum():
```

Comment:
`field.IsEnum()` being treated as integers can be problematic if for any reason we start using `jsonpb` where enums are strings

Reply:
After a bunch of research:
- we can extract enum data from a top level message (e.g. `OsType` enum is described in the `events.OsType` proto descriptor
- currently, enum fields like osLogin are marshalled/unmarshalled as integers. Replacing the type in the schema will only lead to encapsulating integers in strings. If we want to change and extract the potential values in the schema, we will also have to support user-friendly values during marshalling/unmarshalling.

We discussed with @tigrato and will put a TODO there and postpone this change.